agent.go

```diff
@@ -315,7 +290,160 @@ func (a *Agent) watchHealthChecks(updateCh chan api.HealthChecks) {
+		// as well as a filtered list of those nodes which have external-probe set.
+		healthCheckNodes := make(map[string]bool)
+		var pingNodes []*api.Node
+		for i := ourIndex; i < len(externalNodes); i += len(healthyInstances) {
+			node := externalNodes[i]
+			healthCheckNodes[node.Node] = true
+
+			if node.Meta == nil {
+				continue
+			}
+			if _, ok := node.Meta["external-probe"]; ok {
+				pingNodes = append(pingNodes, node)
+			}
+		}
+
+		if len(healthCheckNodes) != prevCount {
+			prevCount = len(healthCheckNodes)
+			a.logger.Printf("[INFO] Now watching %d external nodes", prevCount)
+		}
+
+		if a.watchedNodeFunc != nil {
+			a.watchedNodeFunc(healthCheckNodes, pingNodes)
+		}
+
+		healthNodeCh <- healthCheckNodes
+		pingNodeCh <- pingNodes
+	}
+}
+
+// watchExternalNodes does a watch for external nodes and returns any updates
+// back through nodeCh as a sorted list.
+func (a *Agent) watchExternalNodes(nodeCh chan []*api.Node) {
+	opts := &api.QueryOptions{
+		NodeMeta: a.config.NodeMeta,
+	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	opts = opts.WithContext(ctx)
+	go func() {
+		<-a.shutdownCh
+		cancelFunc()
+	}()
+
+	firstRun := make(chan struct{}, 1)
+	firstRun <- struct{}{}
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case <-firstRun:
+			// Skip the wait on the first run.
+		case <-time.After(retryTime):
+			// Sleep here to limit how much load we put on the Consul servers.
+		}
+
+		// Do a blocking query for any external node changes
+		externalNodes, meta, err := a.client.Catalog().Nodes(opts)
+		if err != nil {
+			a.logger.Printf("[WARN] Error getting external node list: %v", err)
+			continue
+		}
+		sort.Slice(externalNodes, func(a, b int) bool {
+			return externalNodes[a].Node < externalNodes[b].Node
+		})
+
+		opts.WaitIndex = meta.LastIndex
+
+		nodeCh <- externalNodes
+	}
+}
+
+// watchExternalNodes does a watch for any ESM instances with the same service tag as
+// this agent and sends any updates back through instanceCh as a sorted list.
+func (a *Agent) watchServiceInstances(instanceCh chan []*api.ServiceEntry) {
+	var opts *api.QueryOptions
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	opts = opts.WithContext(ctx)
+	go func() {
+		<-a.shutdownCh
+		cancelFunc()
+	}()
+
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case <-time.After(retryTime / 10):
+			// Sleep here to limit how much load we put on the Consul servers. We can
+			// wait a lot less than the normal retry time here because the ESM service instance
+			// list is relatively small and cheap to query.
+		}
+
+		healthyInstances, meta, err := a.client.Health().Service(a.config.Service, a.config.Tag, true, opts)
+		if err != nil {
+			a.logger.Printf("[WARN] Error querying for health check info: %v", err)
+			continue
+		}
+		sort.Slice(healthyInstances, func(a, b int) bool {
+			return healthyInstances[a].Service.ID < healthyInstances[b].Service.ID
+		})
+
+		opts.WaitIndex = meta.LastIndex
```

Comment:
We never actually set a `WaitTime` so this will not actually use blocking queries right? Why would be poll every second rather than block?

Reply:
The default wait time is 5 minutes, so we still get a blocking query